### ISSUE TITLE
fails to install because $ENV{"TRAVIS_PERL_VERSION"} is not set

### ISSUE THREAD

**Issue by stuart-little** [ownlifeful/Graph-Undirected-Hamiltonicity#3]
The docs seem to say nothing about this environment variable. What should I set it to? Why is it mandatory?

Here's the log I get from cloning this repo and running `cpanm .` in the main directory:
```
cpanm (App::cpanminus) 1.7044 on perl 5.035000 built for x86_64-linux
Work directory is $HOME/.cpanm/work/1622457557.212652
You have make /usr/bin/make
You have LWP 6.54
You have /usr/bin/tar: tar (GNU tar) 1.34
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
You have /usr/bin/unzip
--> Working on .
Entering $HOME/repos/Graph-Undirected-Hamiltonicity
Checking if you have ExtUtils::Install 1.46 ... Yes (2.20)
Checking if you have Module::Build 0.38 ... Yes (0.4231)
Configuring $HOME/repos/Graph-Undirected-Hamiltonicity
Running Build.PL
Use of uninitialized value in concatenation (.) or string at Build.PL line 18.
Use of uninitialized value in concatenation (.) or string at Build.PL line 19.
Use of uninitialized value $ENV{"TRAVIS_PERL_VERSION"} in concatenation (.) or string at Build.PL line 20.
Use of uninitialized value $ENV{"TRAVIS_PERL_VERSION"} in concatenation (.) or string at Build.PL line 21.
FOUND NO PREFIX
-> N/A
Building and testing .
Can't locate object method "cpanm" via package "Dist::Zilla" (perhaps you forgot to load "Dist::Zilla"?) at ./Build line 1.
-> FAIL Installing . failed. See $HOME/.cpanm/work/1622457557.212652/build.log for details. Retry with --force to force install it.
```

**Comment by ownlifeful** [comment]
You found a genuine bug. It has been fixed. Please pull/clone and try again.
Thank you!

**Comment by stuart-little** [comment]
@ownlifeful This doesn't really fix things..

If I run `perl Build.PL 1`, as [instructed](/home/chirvasitua/.cpanm/work/1622578358.278665/build.log), I get a `Build` file that just says

```
cpanm Dist::Zilla
```
That's because you [ask the process](https://github.com/ownlifeful/Graph-Undirected-Hamiltonicity/blob/6e12c9d71d207e5154569698e02e67778919af5a/Build.PL#L27) to die if it finds no prefix, and it finds no prefix (same issue as before).

Additionally, if I manually run `dzil install` as [intended](https://github.com/ownlifeful/Graph-Undirected-Hamiltonicity/blob/6e12c9d71d207e5154569698e02e67778919af5a/Build.PL#L33), I get one gigantic failure (this being the output of `dzil install --verbose`):
```
[DZ] reading configuration using Dist::Zilla::MVP::Reader::Finder
[@Basic/GatherDir] online, Dist::Zilla::Plugin::GatherDir v6.017
[@Basic/PruneCruft] online, Dist::Zilla::Plugin::PruneCruft v6.017
[@Basic/ManifestSkip] online, Dist::Zilla::Plugin::ManifestSkip v6.017
[@Basic/MetaYAML] online, Dist::Zilla::Plugin::MetaYAML v6.017
[@Basic/License] online, Dist::Zilla::Plugin::License v6.017
[@Basic/Readme] online, Dist::Zilla::Plugin::Readme v6.017
[@Basic/ExtraTests] online, Dist::Zilla::Plugin::ExtraTests v6.017
[@Basic/ExecDir] online, Dist::Zilla::Plugin::ExecDir v6.017
[@Basic/ShareDir] online, Dist::Zilla::Plugin::ShareDir v6.017
[@Basic/MakeMaker] online, Dist::Zilla::Plugin::MakeMaker v6.017
[@Basic/Manifest] online, Dist::Zilla::Plugin::Manifest v6.017
[@Basic/TestRelease] online, Dist::Zilla::Plugin::TestRelease v6.017
[@Basic/ConfirmRelease] online, Dist::Zilla::Plugin::ConfirmRelease v6.017
[@Basic/UploadToCPAN] online, Dist::Zilla::Plugin::UploadToCPAN v6.017
[AutoPrereqs] online, Dist::Zilla::Plugin::AutoPrereqs v6.017
[DZ] building distribution under .build/ysKvMEo7uz for installation
[DZ] beginning to build Graph-Undirected-Hamiltonicity
[@Basic/GatherDir] considering .
[@Basic/GatherDir] considering WOLFRAM.md
[@Basic/GatherDir] considering Build
[@Basic/GatherDir] considering .travis.yml
[@Basic/GatherDir] considering .gitignore
[@Basic/GatherDir] considering README.md
[@Basic/GatherDir] considering dist.ini
[@Basic/GatherDir] considering Changes
[@Basic/GatherDir] considering Build.PL
[@Basic/GatherDir] considering hamilton.ini
[@Basic/GatherDir] considering t
[@Basic/GatherDir] considering graph_is_hamiltonian.t
[@Basic/GatherDir] considering 000-load.t
[@Basic/GatherDir] considering tests
[@Basic/GatherDir] considering generic_tests.t
[@Basic/GatherDir] considering transforms
[@Basic/GatherDir] considering string_to_graph.t
[@Basic/GatherDir] considering add_random_edges.t
[@Basic/GatherDir] considering get_required_graph.t
[@Basic/GatherDir] considering swap_vertices.t
[@Basic/GatherDir] considering delete_non_required_neighbors.t
[@Basic/GatherDir] considering delete_cycle_closing_edges.t
[@Basic/GatherDir] considering get_random_isomorph.t
[@Basic/GatherDir] considering get_common_neighbors.t
[@Basic/GatherDir] considering spoof
[@Basic/GatherDir] considering spoof_random_graph.t
[@Basic/GatherDir] considering spoof_canonical_hamiltonian_graph.t
[@Basic/GatherDir] considering spoof_randomish_graph.t
[@Basic/GatherDir] considering spoof_known_hamiltonian_graph.t
[@Basic/GatherDir] considering bin
[@Basic/GatherDir] considering knights_tour.pl
[@Basic/GatherDir] considering hamilton.pl
[@Basic/GatherDir] considering cgi-bin
[@Basic/GatherDir] considering hc.cgi
[@Basic/GatherDir] considering .build
[@Basic/GatherDir] considering .git
[@Basic/GatherDir] considering lib
[@Basic/GatherDir] considering Graph
[@Basic/GatherDir] considering Undirected
[@Basic/GatherDir] considering Hamiltonicity.pm
[@Basic/GatherDir] considering Hamiltonicity.pod
[@Basic/GatherDir] considering Hamiltonicity
[@Basic/GatherDir] considering Output.pm
[@Basic/GatherDir] considering Transforms.pm
[@Basic/GatherDir] considering Tests.pm
[@Basic/GatherDir] considering Transforms.pod
[@Basic/GatherDir] considering Tests.pod
[@Basic/GatherDir] considering Spoof.pod
[@Basic/GatherDir] considering Wolfram.pm
[@Basic/GatherDir] considering Wolfram.pod
[@Basic/GatherDir] considering Spoof.pm
[@Basic/GatherDir] considering Output.pod
[@Basic/GatherDir] adding file WOLFRAM.md
[@Basic/GatherDir] adding file Build
[@Basic/GatherDir] adding file README.md
[@Basic/GatherDir] adding file dist.ini
[@Basic/GatherDir] adding file Changes
[@Basic/GatherDir] adding file Build.PL
[@Basic/GatherDir] adding file hamilton.ini
[@Basic/GatherDir] adding file t/graph_is_hamiltonian.t
[@Basic/GatherDir] adding file t/000-load.t
[@Basic/GatherDir] adding file t/tests/generic_tests.t
[@Basic/GatherDir] adding file t/transforms/string_to_graph.t
[@Basic/GatherDir] adding file t/transforms/add_random_edges.t
[@Basic/GatherDir] adding file t/transforms/get_required_graph.t
[@Basic/GatherDir] adding file t/transforms/swap_vertices.t
[@Basic/GatherDir] adding file t/transforms/delete_non_required_neighbors.t
[@Basic/GatherDir] adding file t/transforms/delete_cycle_closing_edges.t
[@Basic/GatherDir] adding file t/transforms/get_random_isomorph.t
[@Basic/GatherDir] adding file t/transforms/get_common_neighbors.t
[@Basic/GatherDir] adding file t/spoof/spoof_random_graph.t
[@Basic/GatherDir] adding file t/spoof/spoof_canonical_hamiltonian_graph.t
[@Basic/GatherDir] adding file t/spoof/spoof_randomish_graph.t
[@Basic/GatherDir] adding file t/spoof/spoof_known_hamiltonian_graph.t
[@Basic/GatherDir] adding file bin/knights_tour.pl
[@Basic/GatherDir] adding file bin/hamilton.pl
[@Basic/GatherDir] adding file cgi-bin/hc.cgi
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity.pm
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity.pod
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Output.pm
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Transforms.pm
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Tests.pm
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Transforms.pod
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Tests.pod
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Spoof.pod
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Wolfram.pm
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Wolfram.pod
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Spoof.pm
[@Basic/GatherDir] adding file lib/Graph/Undirected/Hamiltonicity/Output.pod
[@Basic/MetaYAML] adding file META.yml
[@Basic/License] adding file LICENSE
[@Basic/Readme] adding file README
[@Basic/MakeMaker] adding file Makefile.PL
[@Basic/Manifest] adding file MANIFEST
[@Basic/PruneCruft] pruning Build
[@Basic/ManifestSkip] file MANIFEST.SKIP not found
[DZ] guessing dist's main_module is lib/Graph/Undirected/Hamiltonicity.pm
[DZ] dist's main_module is lib/Graph/Undirected/Hamiltonicity.pm
[DZ] extracting distribution abstract from lib/Graph/Undirected/Hamiltonicity.pm
[@Basic/ExtraTests] MANIFEST has 'bytes' encoding, skipping...
[AutoPrereqs] scanning bin/hamilton.pl for runtime prereqs
[AutoPrereqs] scanning bin/knights_tour.pl for runtime prereqs
[AutoPrereqs] scanning lib/Graph/Undirected/Hamiltonicity.pm for runtime prereqs
[AutoPrereqs] scanning lib/Graph/Undirected/Hamiltonicity/Output.pm for runtime prereqs
[AutoPrereqs] scanning lib/Graph/Undirected/Hamiltonicity/Spoof.pm for runtime prereqs
[AutoPrereqs] scanning lib/Graph/Undirected/Hamiltonicity/Tests.pm for runtime prereqs
[AutoPrereqs] scanning lib/Graph/Undirected/Hamiltonicity/Transforms.pm for runtime prereqs
[AutoPrereqs] scanning lib/Graph/Undirected/Hamiltonicity/Wolfram.pm for runtime prereqs
[AutoPrereqs] excluding local packages: bin::hamilton.pl, bin::knights_tour.pl, Graph::Undirected::Hamiltonicity, Graph::Undirected::Hamiltonicity::Output, Graph::Undirected::Hamiltonicity::Spoof, Graph::Undirected::Hamiltonicity::Tests, Graph::Undirected::Hamiltonicity::Transforms, Graph::Undirected::Hamiltonicity::Wolfram
[AutoPrereqs] scanning t/000-load.t for test prereqs
[AutoPrereqs] scanning t/graph_is_hamiltonian.t for test prereqs
[AutoPrereqs] scanning t/spoof/spoof_canonical_hamiltonian_graph.t for test prereqs
[AutoPrereqs] scanning t/spoof/spoof_known_hamiltonian_graph.t for test prereqs
[AutoPrereqs] scanning t/spoof/spoof_random_graph.t for test prereqs
[AutoPrereqs] scanning t/spoof/spoof_randomish_graph.t for test prereqs
[AutoPrereqs] scanning t/tests/generic_tests.t for test prereqs
[AutoPrereqs] scanning t/transforms/add_random_edges.t for test prereqs
[AutoPrereqs] scanning t/transforms/delete_cycle_closing_edges.t for test prereqs
[AutoPrereqs] scanning t/transforms/delete_non_required_neighbors.t for test prereqs
[AutoPrereqs] scanning t/transforms/get_common_neighbors.t for test prereqs
[AutoPrereqs] scanning t/transforms/get_random_isomorph.t for test prereqs
[AutoPrereqs] scanning t/transforms/get_required_graph.t for test prereqs
[AutoPrereqs] scanning t/transforms/string_to_graph.t for test prereqs
[AutoPrereqs] scanning t/transforms/swap_vertices.t for test prereqs
[AutoPrereqs] excluding local packages: bin::hamilton.pl, bin::knights_tour.pl, Graph::Undirected::Hamiltonicity, Graph::Undirected::Hamiltonicity::Output, Graph::Undirected::Hamiltonicity::Spoof, Graph::Undirected::Hamiltonicity::Tests, Graph::Undirected::Hamiltonicity::Transforms, Graph::Undirected::Hamiltonicity::Wolfram, t::000-load.t, 000-load.t, t::graph_is_hamiltonian.t, graph_is_hamiltonian.t, t::spoof::spoof_canonical_hamiltonian_graph.t, spoof::spoof_canonical_hamiltonian_graph.t, t::spoof::spoof_known_hamiltonian_graph.t, spoof::spoof_known_hamiltonian_graph.t, t::spoof::spoof_random_graph.t, spoof::spoof_random_graph.t, t::spoof::spoof_randomish_graph.t, spoof::spoof_randomish_graph.t, t::tests::generic_tests.t, tests::generic_tests.t, t::transforms::add_random_edges.t, transforms::add_random_edges.t, t::transforms::delete_cycle_closing_edges.t, transforms::delete_cycle_closing_edges.t, t::transforms::delete_non_required_neighbors.t, transforms::delete_non_required_neighbors.t, t::transforms::get_common_neighbors.t, transforms::get_common_neighbors.t, t::transforms::get_random_isomorph.t, transforms::get_random_isomorph.t, t::transforms::get_required_graph.t, transforms::get_required_graph.t, t::transforms::string_to_graph.t, transforms::string_to_graph.t, t::transforms::swap_vertices.t, transforms::swap_vertices.t
[AutoPrereqs] excluding local packages: bin::hamilton.pl, bin::knights_tour.pl, Graph::Undirected::Hamiltonicity, Graph::Undirected::Hamiltonicity::Output, Graph::Undirected::Hamiltonicity::Spoof, Graph::Undirected::Hamiltonicity::Tests, Graph::Undirected::Hamiltonicity::Transforms, Graph::Undirected::Hamiltonicity::Wolfram, t::000-load.t, 000-load.t, t::graph_is_hamiltonian.t, graph_is_hamiltonian.t, t::spoof::spoof_canonical_hamiltonian_graph.t, spoof::spoof_canonical_hamiltonian_graph.t, t::spoof::spoof_known_hamiltonian_graph.t, spoof::spoof_known_hamiltonian_graph.t, t::spoof::spoof_random_graph.t, spoof::spoof_random_graph.t, t::spoof::spoof_randomish_graph.t, spoof::spoof_randomish_graph.t, t::tests::generic_tests.t, tests::generic_tests.t, t::transforms::add_random_edges.t, transforms::add_random_edges.t, t::transforms::delete_cycle_closing_edges.t, transforms::delete_cycle_closing_edges.t, t::transforms::delete_non_required_neighbors.t, transforms::delete_non_required_neighbors.t, t::transforms::get_common_neighbors.t, transforms::get_common_neighbors.t, t::transforms::get_random_isomorph.t, transforms::get_random_isomorph.t, t::transforms::get_required_graph.t, transforms::get_required_graph.t, t::transforms::string_to_graph.t, transforms::string_to_graph.t, t::transforms::swap_vertices.t, transforms::swap_vertices.t
[@Basic/MakeMaker] updating contents of Makefile.PL in memory
[DZ] writing Graph-Undirected-Hamiltonicity in .build/ysKvMEo7uz
[DZ] installing via {{["cpanm", "."]}}
--> Working on .
Configuring Graph-Undirected-Hamiltonicity-0.1 ... OK
Building and testing Graph-Undirected-Hamiltonicity-0.1 ... ! Installing . failed. See $HOME/.cpanm/work/1622579011.279122/build.log for details. Retry with --force to force install it.
FAIL
[DZ] error running {{["cpanm", "."]}} at inline delegation in Dist::Zilla for logger->log_fatal (attribute declared in $HOME/.perlbrew/libs/perl-5.35.0@535/lib/perl5/Dist/Zilla.pm at line 773) line 18.
[DZ] error running {{["cpanm", "."]}}
[DZ] install failed, left failed dist in place at .build/ysKvMEo7uz
```
Finally, If I consult the `cpanm` log I'm sent to, I get 
```
cpanm (App::cpanminus) 1.7044 on perl 5.035000 built for x86_64-linux
Work directory is $HOME/.cpanm/work/1622579011.279122
You have make /usr/bin/make
You have LWP 6.54
You have /usr/bin/tar: tar (GNU tar) 1.34
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
You have /usr/bin/unzip
--> Working on .
Entering $HOME/repos/Graph-Undirected-Hamiltonicity/.build/ysKvMEo7uz
Checking configure dependencies from META.yml
Checking if you have ExtUtils::MakeMaker 6.58 ... Yes (7.62)
Configuring Graph-Undirected-Hamiltonicity-0.1
Running Build.PL
FOUND NO PREFIX
Running Makefile.PL
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for Graph::Undirected::Hamiltonicity
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have Test::More 0 ... Yes (1.302185)
Checking if you have LWP::UserAgent 0 ... Yes (6.54)
Checking if you have Modern::Perl 0 ... Yes (1.20200211)
Checking if you have strict 0 ... Yes (1.12)
Checking if you have Getopt::Long 0 ... Yes (2.52)
Checking if you have Carp 0 ... Yes (1.52)
Checking if you have Exporter 0 ... Yes (5.76)
Checking if you have lib 0 ... Yes (0.65)
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.62)
Checking if you have Test::Exception 0 ... Yes (0.43)
Checking if you have Config::INI::Reader 0 ... Yes (0.025)
Checking if you have Graph::Undirected 0 ... Yes (undef)
Building and testing Graph-Undirected-Hamiltonicity-0.1
Can't locate object method "cpanm" via package "Dist::Zilla" (perhaps you forgot to load "Dist::Zilla"?) at ./Build line 1.
-> FAIL Installing . failed. See $HOME/.cpanm/work/1622579011.279122/build.log for details. Retry with --force to force install it.
```
In short:
```
Can't locate object method "cpanm" via package "Dist::Zilla" (perhaps you forgot to load "Dist::Zilla"?) at ./Build line 1.
```
So `cpanm` gets confused by that `cpanm Dist::Zilla` you insert into the `./Build` file.

**Comment by stuart-little** [comment]
@ownlifeful Incidentally, I am not familiar with the `CPAN` module-creation process, so please forgive me if this is off. I have just followed [this](https://tynovsky.github.io/cpan-tutorials/03-make-your-own-distribution.html) guide to compare the results with your setup.

It seems to me the build, as you set it up, is off: the `./Build` file is not intended to be a shell script, as it will be in your case if I run `perl Build.PL 1`. Rather, it will itself be a `perl` script.

I've prepared an example, following that guide I linked:
- download [this](https://www.dropbox.com/s/sxwawnwooqjydcq/ex.tar.gz?dl=0) `tar.gz` of a  bare-bones distro
- unpack it and `cd` into it
- run `cpanm .`

It will install a module called `Acme::Hola` (that doesn't do much at all). Upon doing either that or just plain `perl Build.PL`, you can also examine the `./Build` file generated. On my phone it reads
```
#!/data/data/com.termux/files/usr/bin/perl

use Module::Build::Tiny;
Build();
```
So it's again a `perl` script, not a bunch of shell commands.

**Comment by stuart-little** [comment]
OK, here's what actually worked for me, for the purpose of installing:
- I cloned this repo;
- I deleted `Build.PL`, as per [this](http://dzil.org/tutorial/convert-dist.html) `Dist::Zilla` walkthrough: the whole point of `Dist::Zilla`, apparently, is to spare module authors having to fiddle with either `Makefile.PL` or `Build.PL` (among other things);
- I ran `dzil build` inside the repo directory;
- this produces a `Graph-Undirected-Hamiltonicity-0.1/` subdirectory and a `.tar.gz` version thereof;
- I moved inside that subdirectory (that now contains a `Makefile.PL` generated by `Dist::Zilla`) and ran `cpanm .`.

That's it: this installs `Graph::Undirected::Hamiltonicity` properly, so `cpanm` knows about it, as does `perldoc`.

I think the installation instructions should be amended to reflect the above..

**Comment by ownlifeful** [comment]
Wow! Thanks so much for your effort. It's highly educational too. At least I learned some new things.

Also, I see you're using perl 5.035
That's a problem, because I'm using a system variable ( $^V ) to get the version of perl. That variable was introduced in perl 5.6.0

cpanm (App::cpanminus) 1.7044 on perl 5.035000 built for x86_64-linux

I should add a directive to require perl 5.6.0 or above.
Thanks so much for your patient debugging.

**Comment by stuart-little** [comment]
> Also, I see you're using perl 5.035
> That's a problem, because I'm using a system variable ( $^V ) to get the version of perl. That variable was introduced in perl 5.6.0

This must be another misunderstanding :): `5.035` isn't *lower* than `5.6.0`, it's *higher*. They have different notation styles, whereby `5.035 = 5.35.0`. 

I use [perlbrew](https://perlbrew.pl/) to compile and install the latest `perl`s, and the latest available on there right now is `5.35.0`. So no worries, `$^V` works fine:
```
$ perl -E 'say $^V'
v5.35.0
```
That wasn't the problem.

**Comment by ownlifeful** [comment]
@stuart-little 

Learned something new again ( the Perl version numbers ).

Incidentally, the TRAVIS_PERL_VERSION , and Build.PL are used by travis-ci.org and travis-ci.com ( Travis-CI is a Continuous Integration service ). The builds have been working fine as of a few commits ago. Running Build.PL generates Build which is supposed to be a shell script ( according to Travis-CI docs ).

It was working fine. Then travis-ci.org ( the free service ) got discontinued, and I've had to migrate to travis-ci.com ( the commercial service with a limited free plan ) and something broke.

I'll fix this issue shortly and test the installation myself. Please stand by.

**Comment by ownlifeful** [comment]
@stuart-little I use perlbrew myself, but was unaware that the latest bleeding edge perl was v5.35.0
Will give that a try. However, I will support the module up to the latest stable perl which is still v5.34.0

In the meanwhile, I'm curious -- does it build correctly on v5.34.0 ? Please try that, if you like.

**Comment by stuart-little** [comment]
I don't have `5.34` anywhere, but I do have

- `5.35.0` on a couple of laptops 
- and `5.32.1` [on my phone](https://termux.com/).

There's exactly no difference in behavior between these two. Why would there be? As reported, `cpanm` chokes on the `Build` shell script because it simply isn't expecting such a file.

So I don't think the `perl` version has much to do with this at all: it's not like it'll 

(a) *not* work on `5.32.1`;
(b) work on `5.34`, and
(c) go back to not working on `5.35`.

**Comment by stuart-little** [comment]
By the way, that process I described, of installing? You can simplify it: just
- remove `Build.PL`
- run `dzil install`
So anyway, the problem seems to be that in its present state, the distro is not packaged in a format that `cpanm` knows how to handle.

Perhaps that `Build` shell script is what `travis` wants (never used it), but in handling the distro with `Dist::Zilla` (with, say, `dzil build` or `dzil install`) the `Build.PL` is just in the way: `cpanm` sees it and tries to process it, failing.

An alternative would be to rename your `Build.PL` to something that's not common in `CPAN`-module-building. Call it, say, `BuildMaker.PL`. Then, having also fixed so that it doesn't die prematurely if finding to prefix, this should work:
- perl BuildMaker.PL 1
- ./Build

It would be very close to your initial install instructions, and work..

[Here's a patch](https://termbin.com/z278) that worked for me when applied to the current repo, if you care to apply it (or something like it). Those changes were made after renaming `Build.PL` to `BuildMaker.PL`, so it's a total of 2 commits.

**Comment by stuart-little** [comment]
RE: Travis:

I don't think this repo is configured properly for that either, despite the fact that you say it used to work. I am looking straight at the [Perl-specific docs](https://docs.travis-ci.com/user/languages/perl/), which in turn give [examples](https://docs.travis-ci.com/user/languages/perl/#examples) of distros using `Travis`.

Take say [leto/math-primality](https://github.com/leto/math--primality). If I clone that repo, `cd` into it and run `perl Build.PL`, the `Build` file it produces is a `perl` (not shell) script.

So I do not believe that `Travis` *requires* the present setup. Could you point me to the precise spot in the docs that gabe that impression?

**Comment by ownlifeful** [comment]
You're right. I just checked the latest docs and Build just has to be an executable script. Could be bash / perl / any other language.

I'm beginning to think that using Dist::Zilla is perhaps not the best solution. This codebase used to utilize ExtUtils::MakeMaker , like most Perl modules. But installing dependencies under that system was a bit unreliable. With Dist::Zilla, all dependencies are included, so the app i s easy to deploy.

I am thinking of bundling this project using Carton or Docker or some other container solution where the dependencies could be included in a reliable and robust manner. Any thoughts / recommendations?

Also, I'm having trouble migrating from travis-ci.org to travis-ci.com Now that Travis is not entirely free, I might have to look at a different CI solution. GitHub Actions might provide a better way to run tests. Any suggestions?

Thanks a bunch!

**Comment by stuart-little** [comment]
Well, I know nothing about either of the two subjects (CI and CPAN publishing), so please take all of this with a grain of salt. All I'm doing here is poking at this with a stick.

Ditching `Dist::Zilla` doesn't seem like a good choice: why would you anyway? The problem isn't there, but rather with the way the project was put together (with that `Build.PL` doing double duty, as a shell-script producer and a hook for `cpanm`). In fact, as you say, `Dist::Zilla` make your life much easier. 

Here's what I tried just now: 
- installed `perl v5.34.0` with `perlbrew` cleanly, to start from scratch (no modules, etc.);
- installed `Dist::Zilla`  
- in your repo, *without* that pesky `Build.PL`, ran 
```
dzil build
```
- moved into the `Graph-Undirected-Hamiltonicity-0.1` subdirectory that produced
- ran `cpanm .`

That worked fine! So this means that you can run `dzil build` at your end, and then only publish the contents of  `Graph-Undirected-Hamiltonicity-<version>` on `GitHub` (not the whole thing). Then your users can simply clone your repo and run `cpanm .`. `cpanm` will find the dependencies and install them, etc. (it's what happened in my case). 

Now, producing `Graph-Undirected-Hamiltonicity-<version>` from the `dist.ini` config with `dzil` is a different matter, and you can decide how to automate that separately. Besides, you wouldn't want/expect your users to have `Dist::Zilla` anyway: that's for package *authors* (as the first line of [the homepage](http://dzil.org/) makes clear).

**Comment by ownlifeful** [comment]
@stuart-little 

The main reason I was thinking of ditching Dist::Zilla is because the user has to download an entire CPAN distro just to meet dependencies. That's not a good tradeoff.

Also, I've applied your patch, and it tries to build Graph-Undirected-Hamiltonicity-0.1
and fails. I don't know why it got the idea to append the version number at the end. It should just build Graph-Undirected-Hamiltonicity

**Comment by stuart-little** [comment]
> @stuart-little
> 
> The main reason I was thinking of ditching Dist::Zilla is because the user has to download an entire CPAN distro just to meet dependencies. That's not a good tradeoff.
> 

Which distro? `Dist::Zilla`? As I said, the user is not expected to download and install it. That's for you, the author. 

> Also, I've applied your patch, and it tries to build Graph-Undirected-Hamiltonicity-0.1
> and fails. I don't know why it got the idea to append the version number at the end. It should just build Graph-Undirected-Hamiltonicity

I applied the patch too, to your master branch, after renaming `Build.PL` to `BuildMaker.PL`. 
```
perl BuildMaker.PL 1 && ./Build
```
works fine for me and installs `Graph::Undirected::Hamiltonicity`, so I don't know what that's all about. 

After I run the first command, `perl BuildMaker.PL 1`, the `Build` file it generates looks like this:
```
$ cat ./Build
---
cpanm Dist::Zilla
$(which dzil) authordeps | cpanm
$(which dzil) listdeps | cpanm
$(which dzil) install
```
It "got the idea" to append the version because you specified `version = 0.1` right in the [dist.ini file](https://github.com/ownlifeful/Graph-Undirected-Hamiltonicity/blob/7bdcf7f01c879cd87d9992cbd8c78f6f8306be80/dist.ini#L3). That's what `Dist::Zilla` does: it reads that file and follows the directives. 

If you run `dzil build` in your root directory, you'll see the following output in the terminal:
```
[DZ] beginning to build Graph-Undirected-Hamiltonicity
[DZ] guessing dist's main_module is lib/Graph/Undirected/Hamiltonicity.pm
[DZ] writing Graph-Undirected-Hamiltonicity in Graph-Undirected-Hamiltonicity-0.1
[DZ] building archive with Archive::Tar; install Archive::Tar::Wrapper 0.15 or newer for improved speed
[DZ] writing archive to Graph-Undirected-Hamiltonicity-0.1.tar.gz
[DZ] built in Graph-Undirected-Hamiltonicity-0.1
``` 
So as you can see, it's `Dist::Zilla` that's appending the version number, because that's just how it works. This has nothing to do with the patch.

**Comment by stuart-little** [comment]
Or even more concretely: [here is](https://www.dropbox.com/s/5bukxprfceje2uq/Graph-Undirected-Hamiltonicity-0.1.tar.gz?dl=0) a tar-gzipped copy of the directory I created by simply running `dzil build` in your repo (after removing `Build.PL`, which is not needed for this; forget the patch).

This `tar.gz` is ready to install, for anyone: if you run `cpanm <path-to-the-tar.gz>` then `cpanm` will automatically retrieve the dependencies. 

These dependencies do *not* include `Dist::Zilla`: I used that at my end when I ran `dzil build`, but the package *user* doesn't need it. 

So what you could do is use `Dist::Zilla` on your own to generate this directory, and then publish *it* both on `CPAN` and here.

**Comment by ownlifeful** [comment]
@stuart-little , Ok, I've made some changes ( removed ./Build.PL; added ./Build ) and published revised Installation instructions. Also, I've been able to install & run the module ( demo: bin/hamilton.pl ) successfully on v5.34.0 & v5.35.0
It should be fixed now. Please let me know if it works for you now.

**Comment by stuart-little** [comment]
Yes, following the installation instructions in `README.md` now works; thanks!

You're still making your user download `Dist::Zilla` though, which is what you said you *didn't* want to do :). But OK.. So I suppose it's safe to close this, as the issue is technically no longer operative.

## An epilogue

In case you still want to streamline this (no longer requiring users to download `Dist::Zilla`)`; it was also educational for me a number of different ways:

[Math::Primality](https://github.com/leto/math--primality) gives an example of what a repo should look like if you want it all:

- a `Build.PL` for `Travis`;
- *and* a `dist.ini` for `Dist::Zilla`.

You can even see a [comment](https://github.com/leto/math--primality/blob/f5c793818f8b6bcb2936f54e5aa5ee8e49a5b2c9/dist.ini#L16) in their `dist.ini` to the effect that `Build.PL` is not needed by `Dist::Zilla`, but they're keeping it anyway to be able to work on it without `dzil`.

With that repo, *both* approaches work for me:

(a) I can simply run `cpanm .` in there and I'll get the package;
(b) but I can also run `dzil build` to have `Dist::Zilla` build it for me.

The latter fails for [unrelated reasons](https://github.com/rjbs/Dist-Zilla/commit/67323109807f444eecb15b4f2a40c9502328bad5) (it got broken by newer versions of `Dist::Zilla`), but I can [patch](https://www.dropbox.com/s/4zjx43rtwq76493/math-primality.diff?dl=0) their `dist.ini` to make `dzil build` work.

Anyway, the point is the repo can be organized in such a manner that 

- you, as an author, can run `dzil build` in there to build the package for publishing;
- while the user can simply run`cpanm .` to install without needing `Dist::Zilla`.

**Comment by ownlifeful** [comment]
Thanks for being interested in this one-off project. It doesn't get cloned & tested very often. I'm learning so much from you.

Incidentally, when everything works, Travis-CI is now broken. So I'm trying to get that working as well, and dealing with some frustration. CI is very useful to have!
